src/connections.jl

```diff
@@ -128,6 +149,40 @@ function handle_new_connection(jl_conn::Connection; throw_error::Bool=true)
     return jl_conn
 end
 
+# Perform the connection loop as specified in
+# https://www.postgresql.org/docs/10/libpq-connect.html#LIBPQ-PQCONNECTSTARTPARAMS
+# NOTE: this uses the non-blocking interface but this function does block!
+function _connect_nonblocking(keywords, values, expand_dbname)
+    conn = libpq_c.PQconnectStartParams(keywords, values, expand_dbname)
+    c_state = libpq_c.PQstatus(conn)
+
+    # This is also true when `conn == C_NULL`
+    c_state == libpq_c.CONNECTION_BAD && return conn
+
```

Comment:
Might want to add a debug message saying that we're polling the socket? I feel like this might be a use-case for `debug1`, `debug2`, `debug3` for Memento logging.